sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/VersioningTest.groovy

```diff
@@ -0,0 +1,364 @@
+
+        then:
+        receivedValV1 == null
+        valV2 == receivedValV2
+        valV3 == receivedValV3
+    }
+
+    def "List Blobs with Version"() {
+        given:
+        def blobItemV1 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        def blobItemV2 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+        def blobItemV3 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+
+        when:
+        def blobs = blobContainerClient.listBlobs(new ListBlobsOptions().setDetails(new BlobListDetails().setRetrieveVersions(true)), null)
+
+        then:
+        blobs.size() == 3
+        blobs[0].getVersionId() == blobItemV1.getVersionId()
+        blobs[1].getVersionId() == blobItemV2.getVersionId()
+        blobs[2].getVersionId() == blobItemV3.getVersionId()
+        blobs[0].isCurrentVersion() == null
+        blobs[1].isCurrentVersion() == null
+        blobs[2].isCurrentVersion()
+    }
+
+    def "List Blobs without Version"() {
+        given:
+        blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+        def blobItemV3 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize, true)
+
+        when:
+        def blobs = blobContainerClient.listBlobs(new ListBlobsOptions().setDetails(new BlobListDetails().setRetrieveVersions(false)), null)
+
+        then:
+        blobs.size() == 1
+        blobs[0].getVersionId() == blobItemV3.getVersionId()
+    }
+
+    def "Begin Copy Blobs with Version"() {
+        given:
+        def blobItemV1 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        def sourceBlob = blobContainerClient.getBlobClient(generateBlobName())
+        sourceBlob.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+
+        when:
+        def pooler = blobClient.beginCopy(sourceBlob.getBlobUrl(), Duration.ofSeconds(1))
+        def copyInfo = pooler.waitForCompletion().getValue()
+
+        then:
+        copyInfo.getVersionId() != null
+        copyInfo.getVersionId() != blobItemV1.getVersionId()
+    }
+
+    def "Copy From Url Blobs with Version"() {
+        given:
+        blobContainerClient.setAccessPolicy(PublicAccessType.CONTAINER, null)
+        def blobItemV1 = blobClient.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+        def sourceBlob = blobContainerClient.getBlobClient(generateBlobName())
+        sourceBlob.getBlockBlobClient().upload(defaultInputStream.get(), defaultDataSize)
+
+        when:
+        def response = blobClient.copyFromUrlWithResponse(sourceBlob.getBlobUrl(), null, null, null, null, null, Context.NONE)
```

Comment:
It may also be worth testing copying from a non-current version. Or at least testing getBlobUrl and ensuring that the version query parameter is present

Reply:
Sounds good. I'll add test case for getBlobUrl.